src/com/amazon/corretto/crypto/provider/AesKeyWrapSpi.java

```diff
@@ -0,0 +1,227 @@
+        implInit(opmode, key, random);
+    }
+
+    private void implInit(int opmode, Key key, SecureRandom random)
+            throws InvalidKeyException, InvalidAlgorithmParameterException {
+        if (opmode != Cipher.UNWRAP_MODE && opmode != Cipher.WRAP_MODE) {
+            throw new UnsupportedOperationException("Cipher only supports un/wrap modes");
+        }
+        if (key == null) {
+            throw new InvalidKeyException("Null key");
+        }
+        if (key != jceKey) {
+            if (!(key instanceof  SecretKey)) {
+                throw new InvalidKeyException("Need a SecretKey");
+            }
+            if (!"RAW".equalsIgnoreCase(key.getFormat())) {
+                throw new InvalidKeyException("Need a raw format key");
+            }
+            if (!"AES".equalsIgnoreCase(key.getAlgorithm())) {
+                throw new InvalidKeyException("Expected an AES key");
+            }
+            if (keyBytes != null) {
+                Arrays.fill(keyBytes, (byte) 0);
+                keyBytes = null;
+            }
+            keyBytes = key.getEncoded();
+            if (keyBytes == null) {
+                throw new InvalidKeyException("Key doesn't support encoding");
+            }
+            jceKey = (SecretKey) key;
+        }
+        this.opmode = opmode;
+        // TODO [childw] do something with |random| here?
+    }
+
+    @Override
+    protected byte[] engineWrap(final Key key) throws IllegalBlockSizeException, InvalidKeyException {
+        if (opmode != Cipher.WRAP_MODE || keyBytes == null) {
+            throw new IllegalStateException("Cipher must be init'd in WRAP_MODE");
+        }
+        try {
+            final byte[] encoded = Utils.encodeForWrapping(provider, key);
+            final byte[] wrappedKey = new byte[4096];
+            int wrappedKeyLen = wrapKey(keyBytes, encoded, wrappedKey);
+            final byte[] wrappedKeyTruncated = new byte[wrappedKeyLen];
+            System.arraycopy(wrappedKey, 0, wrappedKeyTruncated, 0, wrappedKeyLen);
+            return wrappedKeyTruncated;
+        // TODO [childw] make this exception more specific, perhaps BadPaddingException?
+        } catch (final Exception ex) {
+            throw new InvalidKeyException("Wrapping failed", ex);
+        }
+    }
+
+    @Override
+    protected Key engineUnwrap(final byte[] wrappedKey, final String wrappedKeyAlgorithm, final int wrappedKeyType)
+            throws InvalidKeyException, NoSuchAlgorithmException {
+        if (opmode != Cipher.UNWRAP_MODE || keyBytes == null) {
+            throw new IllegalStateException("Cipher must be init'd in WRAP_MODE");
+        }
+        try {
+            final byte[] unwrappedKey = new byte[wrappedKey.length];
+            int unwrappedKeyLen = unwrapKey(keyBytes, wrappedKey, unwrappedKey);
+            final byte[] unwrappedKeyTruncated = new byte[unwrappedKeyLen];
+            System.arraycopy(unwrappedKey, 0, unwrappedKeyTruncated, 0, unwrappedKeyLen);
```

Comment:
As before, don't truncate if we happened to get the correct length. Also, do you want to zero the array that copied and not further used?

(Do you also want to zeroize the used array after the new key is constructed?)